### ISSUE TITLE
new ModuleCollection.execute method only writes into namespace temporarily

### ISSUE THREAD

**Issue by barentine** [python-microscopy/python-microscopy#245]
Looks like a bit of a bug with the next execute context management.

I'm not exactly sure where the scoping is happening, but for some reason if I execute a module which adds a key to the namespace on execution, such that before line 612 I have
```
self.namespace
Out[2]: 
{'input': {'x': array([ 1.,  1.,  1., -1., -1., -1.,  1., -1.]),
  'y': array([ 1.,  1., -1.,  1., -1.,  1., -1., -1.]),
  'z': array([ 1., -1.,  1.,  1.,  1., -1., -1., -1.])}}
```
and after it I have
```
self.namespace
Out[3]: 
{'input': {'x': array([ 1.,  1.,  1., -1., -1., -1.,  1., -1.]),
  'y': array([ 1.,  1., -1.,  1., -1.,  1., -1., -1.]),
  'z': array([ 1., -1.,  1.,  1.,  1., -1., -1., -1.])},
 'output': <PYME.IO.tabular.RecArraySource at 0x81c269518>}
```
as soon as the next iteration of m in exec_order kicks over my `self.namespace` goes back to
```
self.namespace
Out[4]: 
{'input': {'x': array([ 1.,  1.,  1., -1., -1., -1.,  1., -1.]),
  'y': array([ 1.,  1., -1.,  1., -1.,  1., -1., -1.]),
  'z': array([ 1., -1.,  1.,  1.,  1., -1., -1., -1.])}}
```
https://github.com/python-microscopy/python-microscopy/blob/6d3282af614650ddd92e114991987dc2131da7fe/PYME/recipes/base.py#L609-L626

**Comment by barentine** [comment]
Discussed this with @David-Baddeley yesterday who was going to have a look at whether it was due to `ModuleBase` being a `HasTraits` and invalidating. May be fixed in commit 091ad92a8527d558cb4bcd720fa5bb838f114278

**Comment by barentine** [comment]
This has been resolved - thanks @David-Baddeley.
